controllers/components/kserve/kserve_support.go

```diff
@@ -0,0 +1,302 @@
+package kserve
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"path"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	componentsv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/components/v1"
+	dsciv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/dscinitialization/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/cluster"
+	odhtypes "github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/deploy"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature/manifest"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature/serverless"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature/servicemesh"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/resources"
+)
+
+var serviceAccounts = map[cluster.Platform][]string{
+	cluster.Unknown:          {odhModelControllerComponentName},
+	cluster.OpenDataHub:      {odhModelControllerComponentName},
+	cluster.ManagedRhods:     {odhModelControllerComponentName},
+	cluster.SelfManagedRhods: {odhModelControllerComponentName},
+}
+
+// clusterRoleAggregationRulePredicate is a watch predicate that can be used with
+// ClusterRoles to ignore the rules field on update if aggregationRule is set.
+var clusterRoleAggregationRulePredicate = predicate.Funcs{
```

Comment:
i am not sure i understand this part: in which example case, this should be return false as not trigger reconcile?

Reply:
I've changed it now to reflect what Luca suggested (and also I've moved it to `pkg/controller/predicates/clusterrole/clusterrole.go`), so it might make more sense now?

If not: essentially, if aggregationRule is set, then even though the resource we apply has an empty ruleset, k8s will populate it with the rules from the other ClusterRoles that are selected by the aggregationRule section. If we reconcile back to an empty rule set, then it will just be flip-flopping continuously between empty and aggregated, so this predicate will return false when `aggregateRule` is set AND the change is only to the `rules`, since the `rules` are more managed by k8s rather than the operator.

Does that make sense?

Reply:
It does make sense to me.

As a follow up improvement, we can try to remove the `rules` field before the resource gets applied in case it is an aggregation and the rules is empty.  But it requires some small experiments and does not need to be done in this pr.

Reply:
so in short, this predicate is to return false:
if old `.spec.aggregationRule `block and the new` .spec.aggregationRule` block are the same.
(regardless,  old has .spec.rules or new has .spec.rules or if both's .spec.rules are the same/or not) ? 

but, why this did not happen on "incubation"?  the old [modelMeshRolePredicates](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/controllers/datasciencecluster/datasciencecluster_controller.go#L426) does not catch `kserve-admin` which is the only one uses aggregationRule

Reply:
> but, why this did not happen on "incubation"

I think this depends on the SSA vs Merge Patch semantic so it may have happened that the object were left untouched.